packages/block-editor/src/components/typewriter/index.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { Component, createRef } from '@wordpress/element';
+import { computeCaretRect, getScrollContainer } from '@wordpress/dom';
+import { withSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+
+class Typewriter extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.ref = createRef();
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.onMouseDown = this.onMouseDown.bind( this );
+		this.onTouchStart = this.onTouchStart.bind( this );
+		this.maintainCaretPositionOnSelectionChange = this.maintainCaretPositionOnSelectionChange.bind( this );
+		this.computeCaretRectOnSelectionChange = this.computeCaretRectOnSelectionChange.bind( this );
+		this.maintainCaretPosition = this.maintainCaretPosition.bind( this );
+		this.computeCaretRect = this.computeCaretRect.bind( this );
+		this.debouncedComputeCaretRect = debounce( this.computeCaretRect, 100 );
+		this.isSelectionEligibleForScroll = this.isSelectionEligibleForScroll.bind( this );
+	}
+
+	componentDidMount() {
+		window.addEventListener( 'scroll', this.debouncedComputeCaretRect, true );
+		window.addEventListener( 'resize', this.debouncedComputeCaretRect, true );
+	}
+
+	componentWillUnmount() {
+		window.removeEventListener( 'scroll', this.debouncedComputeCaretRect, true );
+		window.removeEventListener( 'resize', this.debouncedComputeCaretRect, true );
+		document.removeEventListener( 'selectionchange', this.maintainCaretPositionOnSelectionChange );
+		document.removeEventListener( 'selectionchange', this.computeCaretRectOnSelectionChange );
+		window.cancelAnimationFrame( this.rafId );
+		this.debouncedComputeCaretRect.cancel();
+	}
+
+	computeCaretRect() {
+		if ( this.isSelectionEligibleForScroll() ) {
+			this.caretRect = computeCaretRect();
+		}
+	}
+
+	maintainCaretPositionOnSelectionChange() {
+		document.removeEventListener( 'selectionchange', this.maintainCaretPositionOnSelectionChange );
+		this.maintainCaretPosition();
+	}
+
+	computeCaretRectOnSelectionChange() {
+		document.removeEventListener( 'selectionchange', this.computeCaretRectOnSelectionChange );
+		this.computeCaretRect();
+	}
+
+	isSelectionEligibleForScroll() {
+		return (
+			// There should be one and only one block selected.
+			this.props.selectedBlockClientId &&
+			// The component must contain the selection.
+			this.ref.current.contains( document.activeElement ) &&
+			// The active element must be contenteditable.
+			document.activeElement.getAttribute( 'contenteditable' )
```

Comment:
Technically the attribute value could be `'false'`, which (ironically) is truthy.

https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes#attr-contenteditable

Perhaps instead, use the `isContentEditable` boolean DOM property:

```suggestion
			document.activeElement.isContentEditable
```

https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/isContentEditable

Reply:
I think either could be fine. I wanted to check whether the element has the attribute, while `isContentEditable` is defined for any element. So a child of an editable field is also editable, but wouldn't have the `contenteditable` attribute. In this case either works fine I guess :)